### PR TITLE
Add CORS parsing/building functions

### PR DESCRIPTION
**Pull request by manifest**
Part of the Cowboy's feature [947](https://github.com/ninenines/cowboy/issues/947) "Make Cowboy CORS friendly". If those changes are ok, I will update Cowboy's [PR](https://github.com/ninenines/cowboy/pull/949).

### CODE REVIEW

**Comment by essen** [comment]
On the interface side of things, two comments: http|https should really be <<"http">> and <<"https">> because that's what we'll have on the Cowboy side of things (either are fine technically, just keeping binary because that's what we get and we don't match too often on them).

Then the reference I don't understand.

On the tests: Cowboy will never send input with trailing whitespace. It's cut off beforehand. That's why parse_host does a binary_to_integer directly, because at that point we can only have numbers or the input is invalid and we crash.

On implementation: don't modify parse_host to attempt to reuse it. This most likely made it slower for no good reason. This module repeats itself a lot, there's no attempt at avoiding that, there's a large attempt at making the code as fast as possible instead. I'd expect parse_origin to have much of the same code as parse_host, except with extra scheme and list handling.

**Comment by manifest** [comment]
According to the [RFC](http://tools.ietf.org/html/rfc6454#section-4) we should generate `a fresh globally unique identifier and return that value` in the cases:

- URI does not use a hierarchical element as a naming authority or if the URI is not an absolute URI
- the implementation doesn't support the protocol given by uri-scheme

So, reference represents a guid here.
Also a CORS implementation needs some representation of the origin even we didn't fully recognize it. It needs something that could be compared with an allowed origins list or "&#x2a;". Thus, we're parsing an "origin" header, getting a reference, comparing it with "&#x2a;" and returning the "access-control-allow-origin" header with the value "&#x2a;". If we failed on parsing origin, we couldn't provide resource which is allowed for all.

As for the tests and implementation I got it and will fix that.

**Comment by essen** [comment]
Fair enough. But this should be written somewhere. :-)

**Comment by manifest** [comment]
Ready.

**Comment by essen** [comment]
That looks very good, best PR in a long time. I don't have any further comments. It looks ready to merge, perhaps tomorrow if time allows. Thanks!

**Comment by manifest** [comment]
Thanks! I will make a few more soon. :-)

**Comment by manifest** [comment]
I've just added other commits to this branch.

**Comment by manifest** [comment]
Probably, we should rename these fuctions:

> access_control_**allow**\_{headers,methods}
> parse_access_control_**allow**\_{headers,methods}
> access_control_**expose**\_headers
> parse_access_control_**expose**\_headers

to

> access_control_**allowed**\_{headers,methods}
> parse_access_control_**allowed**\_{headers,methods}
> access_control_**exposed**\_headers
> parse_access_control_**exposed**\_headers

They will sound more natural, I think. Shall we?

**Comment by essen** [comment]
Should be the header name.

**Comment by manifest** [comment]
The PR is ready then. I've also updated the Cowboy's [PR](https://github.com/ninenines/cowboy/pull/949).

**Comment by essen** [comment]
Yeah I will get back to you before the end of the week hopefully. Today is RabbitMQ day. :-) Thanks!

**Comment by manifest** [comment]
Looking forward to it :-)

**Comment by essen** [comment]
I will put a couple notes, if you have time to fix, cool, otherwise just say so and i'll correct in the merge commit.

File: src/cow_http_hd.erl [hunk 1]
```diff
@@ -2087,6 +2172,148 @@ parse_max_forwards_error_test_() ->
...
+%%
+%% We only seek to have legal characters and separate the
+%% host and port values. The number of segments in the host
+%% or the size of each segment is not checked.
+%%
+%% There is no way to distinguish IPv4 addresses from regular
+%% names until the last segment is reached therefore we do not
+%% differentiate them.
+%%
+%% The following valid hosts are currently rejected: IPv6
```
Comment (essen) [inline src/cow_http_hd.erl]: Should put a `@todo` at the beginning of this line to not forget.

File: src/cow_http_hd.erl [hunk 2]
```diff
@@ -3021,3 +3391,9 @@ token_ci(R, Acc, T) -> token_ci_list_sep(R, [T|Acc]).
 token_ci_list_sep(<<>>, Acc) -> lists:reverse(Acc);
 token_ci_list_sep(<< C, R/bits >>, Acc) when ?IS_WS(C) -> token_ci_list_sep(R, Acc);
 token_ci_list_sep(<< $,, R/bits >>, Acc) -> token_ci_list(R, Acc).
+
+join_token_list([]) -> [];
+join_token_list([H|T]) -> join_token_list(T, [H]).
+
+join_token_list([], Acc) -> lists:reverse(Acc);
+join_token_list([H|T], Acc) -> join_token_list(T, [H,<<$,>>|Acc]).
```
Comment (essen) [inline src/cow_http_hd.erl]: This should be <<", ">>; while the whitespace is optional, it's also recommended to put a single space when it helps readability.

**Comment by essen** [comment]
Think that's it.

**Comment by manifest** [comment]
I will have done it by the end of this day.

**Comment by manifest** [comment]
It's ready.

**Comment by essen** [comment]
Thanks. I'll be on it tomorrow.

**Comment by essen** [comment]
Merged, thanks!

**Pull request merged**